modules/com_vtiger_workflow/tasks/CBWatermark.php

```diff
@@ -0,0 +1,172 @@
+ * granted by the License. coreBOS distributed by JPL TSolucio S.L. is distributed in
+ * the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied
+ * warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. Unless required by
+ * applicable law or agreed to in writing, software distributed under the License is
+ * distributed on an "AS IS" BASIS, WITHOUT ANY WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific language governing
+ * permissions and limitations under the License. You may obtain a copy of the License
+ * at <http://corebos.org/documentation/doku.php?id=en:devel:vpl11>
+ *************************************************************************************************
+ *  Author       : JPL TSolucio, S. L.
+ *************************************************************************************************/
+require_once 'modules/com_vtiger_workflow/VTEntityCache.inc';
+require_once 'modules/com_vtiger_workflow/VTWorkflowUtils.php';
+require_once 'include/events/include.inc';
+
+class CBWatermark extends VTTask {
+	public $executeImmediately = true;
+	public $queable = true;
+
+	public function getFieldNames() {
+		return array('imagesvalue', 'imagesx', 'imagesy', 'exptype');
+	}
+
+	public function doTask(&$entity) {
+		global $current_user, $logbg, $from_wf, $currentModule, $root_directory;
+		$from_wf = true;
+		$logbg->debug('> CBWatermark');
+		$hold_ajxaction = isset($_REQUEST['ajxaction']) ? $_REQUEST['ajxaction'] : '';
+		$_REQUEST['ajxaction'] = 'Workflow';
+		if (!empty($this->imagesvalue)) {
+			$watermark = '';
+			if ($this->exptype == 'rawtext') {
+				$watermark = $this->imagesvalue;
+			} elseif ($this->exptype == 'fieldname') {
+				$util = new VTWorkflowUtils();
+				$adminUser = $util->adminUser();
+				$entityCache = new VTEntityCache($adminUser);
+				$fn = new VTSimpleTemplate($this->imagesvalue);
+				$watermark = $fn->render($entityCache, $entity->getId(), [], $entity->WorkflowContext);
+			} else {
+				$parser = new VTExpressionParser(new VTExpressionSpaceFilter(new VTExpressionTokenizer($this->imagesvalue)));
+				$expression = $parser->expression();
+				$exprEvaluater = new VTFieldExpressionEvaluater($expression);
+				$watermark = $exprEvaluater->evaluate($entity);
+			}
+			$waterMarkUrl = $watermark;
+			$waterMarkArr = explode('/', $waterMarkUrl);
+			$waterMarkDotArr = explode('.', $waterMarkUrl);
+			$waterMarkName = $waterMarkArr[count($waterMarkArr) - 1];
+			$waterMarkType = $waterMarkDotArr[count($waterMarkDotArr) - 1];
+			$waterMarkSavedPath = $root_directory . 'storage/';
+			$data = $entity->getData();
+			$mainImageFileName = $data['filename'];
+			$mainImageFileType = $data['filetype'];
+			$mainImageFileSize = $data['filesize'];
+			$mainImageDownloadUrl = $data['_downloadurl'];
+			$mainImagePath = substr($root_directory, 0, strlen($root_directory) - 9) . parse_url($mainImageDownloadUrl)['path'];
```

Comment:
if I understand correctly these last 5 lines, this task only works with the Documents module, not any image field on any module. If this is the case, we should permit this workflow only for the Documents module adding the module to the changeset

https://github.com/tsolucio/corebos/pull/1448/files#diff-c342bcb94dfa45251e6927c5ecebae0927c68861ff30ca9f6de051dd2f2e4eabR28

or modify the task to permit selecting the image field we want to apply the watermark to